src/passes/pass.cpp

```diff
@@ -891,18 +891,48 @@ void PassRunner::runPass(Pass* pass) {
 
 void PassRunner::runPassOnFunction(Pass* pass, Function* func) {
   assert(pass->isFunctionParallel());
+
+  auto passDebug = getPassDebug();
+
+  // Add extra validation logic in pass-debug mode 2. The main logic in
+  // PassRunner::run will work at the module level, and here for a function-
+  // parallel pass we can do the same at the function level: we can print the
+  // function before the pass, run the pass on the function, and then if it
+  // fails to validate we can show an error and print the state right before the
+  // pass broke it.
+  //
+  // Skip nameless passes for this. Anything without a name is an internal
+  // component of some larger pass, and information about it won't be very
+  // useful - leave it to the entire module to fail validation in that case.
+  bool extraFunctionValidation =
+    passDebug == 2 && options.validate && !pass->name.empty();
+  std::stringstream bodyBefore;
+  if (extraFunctionValidation) {
+    bodyBefore << *func->body << '\n';
+  }
+
   // function-parallel passes get a new instance per function
   auto instance = std::unique_ptr<Pass>(pass->create());
   std::unique_ptr<AfterEffectFunctionChecker> checker;
-  if (getPassDebug()) {
+  if (passDebug) {
     checker = std::unique_ptr<AfterEffectFunctionChecker>(
       new AfterEffectFunctionChecker(func));
   }
   instance->runOnFunction(this, wasm, func);
   handleAfterEffects(pass, func);
-  if (getPassDebug()) {
+  if (passDebug) {
     checker->check();
   }
+
+  if (extraFunctionValidation) {
+    if (!WasmValidator().validate(func, *wasm, WasmValidator::Minimal)) {
```

Comment:
What do we do when the validation mode is `WasmValidator::Minimal`?

Reply:
In minimal mode we don't do any global checks. Like we don't check if a call's target exists. Maybe "minimal" could be "local" or "function-level"...?

Reply:
I asked because we don't currently seem to use this `Minimal` value anywhere in the toolchain. (I grepped with `Minimal`)

And why do we skip the `extraFunctionValidation` if it's `MInimal`, given that this is a function validation?

Reply:
Yeah, I guess Minimal is just 0, so it's "nothing specified" (no global validation, on web validation, no quiet validation). Maybe we can just delete it and use 0,  if we think that's clearer.

I don't understand the second question? We only do this extra work if we decided to do `extraFunctionValidation`. That the work is just a non-global validation is separate from that?

Reply:
Oh, I was just wondering why we have that `if` clause here, if `extraFunctionValidation` and minimal validation are orthogonal things.

Reply:
Oh, just that I thought to do this extra minimal validation only when pass-debug=2 (which makes `extraFunctionValidation` true). Mode 2 is much slower than mode 1, and I didn't want to slow down mode 1 here.